### DISCUSSION TITLE
2024-08-02 - How Much Effect Can a Database Index Have - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by bradystroud** [SSWConsulting/SSW.Rules.Content#8960] [Chewing the Fat]
Any success stories with database indexes?
Any stories of database indexes making things worse?

https://ssw.com.au/rules/sql-real-world-indexes

**Comment by Freego1783** [comment]
Not really a DB guy so I don't have any good stories 😬

**Comment by PothieuG** [comment]
Great video Bryden !

I don't have much experience with large databases, but in the instances where I did, everything went smoothly

**Comment by matt-goldman-ssw** [comment]
You think Sentinel performance is bad now? You should have seen the first version where I made the primary key a concatenation of various other fields rather than just using a composite key 🤦

**Comment by TobyChurches** [comment]
Great Video @brydeno !!

I was working on fixing a query in TimePro relating to iterations. It would also look at the Timesheets related to get the latest timesheet (i.e. when it was last used). This was fixed with the help of @jernejk  and @brydeno, and had a massive performance impact. It reduced the logical reads by a factor of 100,000!! (yes - one hundred thousand)

**Comment by leahy268** [comment]
Great Video @brydeno 

I guess I kind of understood how indexes worked already, but that was such a simple analogy of how they work.  Love it..

>Any success stories with database indexes?

No, not really created any large DB's before myself.  Have used some badly performing databases but didn't really have any control over them.

>Any stories of database indexes making things worse?

No, I wouldn't have had that kind of control.

**Comment by MarcusIrmscherSSW** [comment]
I don't have any experience with database indexes, but I imagine it would help a lot with scalability.

**Comment by GabbyWard** [comment]
An interesting read and video @brydeno 📔

I haven't worked on any DBs yet, but I definitely know some prior clients from my Wood days that should have used something like this to shorten their wait times for data exports.

**Comment by jakebayliss** [comment]
Unique indexes have stopped me accidentally inserting duplicated data, so thats another great thing about them :)

**Comment by tkapa** [comment]
There's always a story about how XYZ Index shaved 0.2 seconds off of a query, making a process that previously took 2 hours, take 10 minutes.

Indexing is obviously a powerful tool, but it seems to have either fallen out of style, or isn't taught much anymore these days.

**Comment by ricksu978** [comment]
I have been using .NET's `Guid.NewGuid()` as a value for primary key in the past.
This has caused tremendous index fragmentation which crippled the database performance.

Alternate solution is - let EF Core generate the Guid value

**Comment by ncn-ssw** [comment]
SugarLearning was having issues with the leaderboard database query timing out at the end of last year. Since @jernejk and @Dhruv-0987 improved the indexes in the database, we have seen no further issues.

**Comment by Ben0189** [comment]
Great video, I havent really worked much on database index

**Comment by steven0x51** [comment]
Great video, good analogy of cook book, that explained database index that eveyone would understand.
I haven't worked much on databse project yet.

**Comment by joshbermanssw** [comment]
Have never worked too much database side, nor much thought onto applying indexes to one.

**Comment by edkranz** [comment]
My understanding of indexes likens it to graphics pre-processing — utilising the fact that we typically have lots of storage, and are limited by time/network speeds. Indexes let you shave down time complexity at the cost of space complexity, a typically worthwhile tradeoff.

**Comment by isaaclombardssw** [comment]
I've never had to use indexes practically but allowing a faster search (when the index has been chosen well) is a great tool.

**Comment by Tanya-Leahy** [comment]
Loved the video!

I have worked as an end user of databases but never understood why they ran fast or slow. This explanation was great, @brydeno.

**Comment by bettybondoc** [comment]
Great video! Made the concept easier to understand. ✨

**Comment by Calinator444** [comment]
> Any success stories with database indexes?

I haven't worked on SQL projects large enough to warrant using database indexes

> Any stories of database indexes making things worse?

I'm unaware of how this could be an issue. @bradystroud can we create a rule on this? We don't have one as far as I'm aware.

**Comment by awaltos** [comment]
I have not worked in the DB realm too much so can't share any interesting stories, but I do have plenty of experience using Sir Jamie's indexes. Great analogy - made it very easy to understand the concept of DB indexes.

**Comment by kikibianc** [comment]
Beautiful analogy - comparing it to a book index was very easy to understand.
Sometimes is important to go back to the meaning and first creation of a concept and we'll understand then and there what they are :)

**Comment by chrisschultzssw** [comment]
> Any success stories with database indexes?
> Any stories of database indexes making things worse?

N/A, not a dev :).

Great explanation though! 🧑🏼‍🍳

**Comment by jeoffreyfischer** [comment]
Now I understand the power of indexing - thanks for a well-explained video Bryden :)

**Comment by wicksipedia** [comment]
I've seen the index foot guns where people have indexed all the columns in a table to "make it faster" 🤦‍♂️

More people should know about https://learn.microsoft.com/en-us/sql/relational-databases/indexes/create-indexes-with-included-columns?view=sql-server-ver16 and how to create these with EF code first

**Comment by duanxinhuan** [comment]
Indexing is good.

**Comment by christoment** [comment]
I am not really a DB person so not much story from my side :)

Curious if using auto performance tuning is a good trade-off for developer team to kickstart project and only manually fine-tune it when needed.

**Comment by AntPolkanov** [comment]
Great explanation, thanks for the video!
When I was looking for my first job, interview questions about DB indexes were very popular 😄

**Comment by Dhruv-0987** [comment]
Not really haven't really played with databases much yet, but I did put an index on a table in SugarLearning when the endpoint for the leaderboard was very slow

**Comment by stephanfakossw** [comment]
N/a - not a dev...Easy to follow video.

**Comment by amankumarrr** [comment]
Recently, I dealt with an issue involving a large dataset (400GB+) for search criteria. I wanted to use indexes to optimize the process, but without access to the data or the ability to simulate it locally, I was concerned about potential impacts on other operations like insertion and updating. Due to time constraints, we couldn't implement the indexes. However, with JK's help, I was able to understand the problem better, and we managed to segregate the queries to make our search much faster.

**Comment by bahjatalaadelssw** [comment]
> Any success stories with database indexes?
> Any stories of database indexes making things worse?
Unfortunately no, but Bryden is a great storyteller, it was a pleasure putting this video together.

**Comment by drwharris** [comment]
Interesting how many devs dont care for the DB side. Ive spent many an engagement before SSW doing DB tuning. SQL server makes this process a pleasure compared to a lot of others.

One of the biggest problems I find is people indexing for the wrong data profile, eg heavy read when its predominately insert.

**Comment by will-truscott-ssw** [comment]
Any success stories with database indexes?
`Yep! Without a decent auto-indexer, most of the projects I'd worked on would have taken far too long to load! Though I did end up having to create a local cache in the most relevant case which made the process a tad irrelevant.`
Any stories of database indexes making things worse?
`From what I understand, if you're writing to the db super frequently, indexes will likely never be usable and just eat up time. In these cases, I've feel database architecture would be far more important as it means you can isolate your less frequently read data and get a usable index back.`

**Comment by MattParkerDev** [comment]
> Any success stories with database indexes?

CRM doesn't let us manage these, so joy of joys.

> Any stories of database indexes making things worse?

I do not have any :)

**Comment by Aibono1225** [comment]
I don't have many good stories since I haven't worked much on the DB side

**Comment by jimmidier** [comment]
One of the difficulties of choosing indexes is to find out which fields are commonly queried. The other one is to balance the usage of indexes with redundant columns. Sometimes a redundant column can avoid join hell and solve performance problem right away, no need to spend hours on execution plan just to end up adding too many indexes.

**Comment by jerryluossw** [comment]
> Any success stories with database indexes?

Have reduced the query time of a complex query from 64s to 500ms through indexes.

> Any stories of database indexes making things worse?

After adding the indexes, the query time in the Staging reduced to 100ms. However, in the Prod, the query time increased instead. The reason is that the query optimizer in the Prod made a different decision compared to the Staging, which is very strange.

**Comment by AnkyChe** [comment]
I'm not a dev, but I find the video easy to understand!

**Comment by tino-liu** [comment]
I don't have any success stories with database. But I am curious on how the drawbacks are for using indexing.. One thing I read is that it could potentially increase the disk space required.

**Comment by babakamyljanovssw** [comment]
As @ncn-ssw and @Dhruv-0987 mentioned, SugarLearning had issue with leaderboard database query.

**Comment by KristenHu** [comment]
I don't have much experience with DB so it's hard to provide any stories.

**Comment by bradystroud** [comment]
They seem simpler than expected.
Interested in learning the code first approach to indexing.

I wonder if the concept is applicable to storing content in markdown.

**Comment by danielmackay** [comment]
> Any success stories with database indexes?
Not really.

> Any stories of database indexes making things worse?
Not really.

**Comment by PennyWalker** [comment]
Reading the above dev responses I've learned a lot from other people's experience. Was great to see Toby's fix on TimePro :)

**Comment by uly1** [comment]
Well done Byrdyen... super clear explanaton!

**Comment by vladislav-kir** [comment]
Great video by Bryden! 

I have not previously created database indexes myself, but saw in action how they majorly improved the performance of the database queries.

**Comment by Hona** [comment]
Thanks Bryden - I've been watching Brent Ozar's Think like the Engine series slowly... :)

**Comment by stephanfakossw** [comment]
[https://outlook-1.cdn.office.net/assets/reaction/heart.png]    Stephan Fako [SSW] reacted to your message:




[cid:ssw-circle(1)_4c7259a6-23f6-49e5-b933-230ca304562e.png]

Stephan Fako
SSW Senior Accountant
Stalk: SSW People<https://www.ssw.com.au/people/Stephan-Fako?utm_source=emailsignature>   |   Walk: SSW Brisbane<https://www.ssw.com.au/ssw/Company/Offices/?utm_source=emailsignature>   |   Talk: +61 451 901 807<https://www.ssw.com.au/ssw/Company/ContactUs.aspx>
[cid:youtube-icon_b8b5cda9-a973-4973-8763-cf3d4922e060.png] <https://www.youtube.com/c/SSWTV>    [cid:linkedin-icon_b153ac9a-eb66-47d9-9c51-ad20b42b8a0a.png]  <https://www.linkedin.com/company/ssw/>    [cid:x-logo_3eb6d0d3-24fb-487d-a4a7-2e30cddc1ebf.png]  <https://x.com/SSW_TV>    [cid:teams-icon_552803c2-9c44-4267-b563-7fe1a7a0e589.png] ***@***.***>


        TinaCMS, the #1 CMS on GitHub, is now part of the SSW family<https://adamcogan.com/2024/05/15/ssw-purchases-tinacms?utm_source=emailsignature>

________________________________

**Comment by ozair-ashfaque** [comment]
Thanks, @brydeno for the video. Great explanation

No success or bad stories yet

**Comment by jernejk** [comment]
Great video @brydeno! While I haven't yet the opportunity to vastly improve performance with indexes yet (it turned out that indexes can't fix exceptionally bad queries), I'm looking forward to a success story using indexes. :)

Last year, I started to add indexes to my project in addition to optimising my queries and I rarely have cases where I need to do any further optimization unless the data scale is truly massive (100k+).

Generic tips for others for DBs in general:

- Never use lazy loading, it will backfire when you least expect it
- Select only the column you need
- Use pagination on the API side (doing it only in the browser doesn't count!)
- Relational DBs: Reduce the amount of join you do in queries
- Cache or fetch separately small data that repeat often (for instance, categories, languages, etc.)
   - For example, instead of fetching 500 books + category names, you can fetch 500 books, create a list of unique category IDs used by 500 books and fetch them separately. Query splitting can sometimes 100x speed up when the amount of data requested increases.
- Changing the table you start in `FROM` section may result in better performance/easier query
   - For instance, fetching all books from an author:
     - Most would do `SELECT b.Id [BookId], a.Name [AuthorName], b.Title [BookTitle] FROM Authors a JOIN Books b ON a.Id = b.AuthorId WHERE a.Id = @AuthorId`
     - Potentially better query `SELECT b.Id [BookId], a.Name [AuthorName], b.Title [BookTitle] FROM Books b JOIN Author a ON a.Id = b.AuthorId WHERE b.AuthorId = @AuthorId`
     - EF Core tip: You can use `AsSplitQuery()` to fetch author separately and potentially improve performance, even more, when there are a lot of books


EF Core specific ones:

- Use `TagWith` or if available `TagWithContext` for your queries, it will add that content as a comment in SQL query for easier debugging
- Use interceptors for entity auditing over manually setting `CreatedOn`, `CreatedBy`, etc.
- When you need a lot of joins or duplicate data (like a category for the book), try using `AsSplitQuery()`
- EF Core is a lot more powerful than most people believe/know

**Comment by SamBWagner** [comment]
Nup. Not really a database doooood. I did relational DB courses in uni but that was eons ago.
